include/nbl/builtin/hlsl/workgroup/bitonic_sort.hlsl

```diff
@@ -0,0 +1,253 @@
+#define NBL_BUILTIN_HLSL_WORKGROUP_BITONIC_SORT_INCLUDED
+#include "nbl/builtin/hlsl/bitonic_sort/common.hlsl"
+#include "nbl/builtin/hlsl/memory_accessor.hlsl"
+#include "nbl/builtin/hlsl/functional.hlsl"
+#include "nbl/builtin/hlsl/subgroup/bitonic_sort.hlsl"
+#include "nbl/builtin/hlsl/bit.hlsl"
+#include "nbl/builtin/hlsl/workgroup/shuffle.hlsl"
+#include "nbl/builtin/hlsl/workgroup/basic.hlsl"
+#include "nbl/builtin/hlsl/concepts/accessors/bitonic_sort.hlsl"
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace workgroup
+{
+namespace bitonic_sort
+{
+
+template<uint16_t _ElementsPerInvocationLog2, uint16_t _WorkgroupSizeLog2, typename KeyType, typename ValueType, typename Comparator = less<KeyType> NBL_PRIMARY_REQUIRES(_ElementsPerInvocationLog2 >= 1 && _WorkgroupSizeLog2 >= 5)
+struct bitonic_sort_config
+{
+    using key_t = KeyType;
+    using value_t = ValueType;
+    using comparator_t = Comparator;
+    NBL_CONSTEXPR_STATIC_INLINE uint16_t ElementsPerInvocationLog2 = _ElementsPerInvocationLog2;
+    NBL_CONSTEXPR_STATIC_INLINE uint16_t WorkgroupSizeLog2 = _WorkgroupSizeLog2;
+
+    NBL_CONSTEXPR_STATIC_INLINE uint32_t ElementsPerInvocation = 1u << ElementsPerInvocationLog2;
+    NBL_CONSTEXPR_STATIC_INLINE uint32_t WorkgroupSize = 1u << WorkgroupSizeLog2;
+    NBL_CONSTEXPR_STATIC_INLINE uint32_t SharedmemDWORDs = sizeof(pair<key_t, value_t>) / sizeof(uint32_t) * WorkgroupSize;
+
+};
+}
+
+template<typename Config, class device_capabilities = void>
+struct BitonicSort;
+
+// ==================== ElementsPerThreadLog2 = 1 Specialization (No Virtual Threading) ====================
+// This handles arrays of size WorkgroupSize * 2 using subgroup + workgroup operations
+template<uint16_t WorkgroupSizeLog2, typename KeyType, typename ValueType, typename Comparator, class device_capabilities>
+struct BitonicSort<bitonic_sort::bitonic_sort_config<1, WorkgroupSizeLog2, KeyType, ValueType, Comparator>, device_capabilities>
+{
+    using config_t = bitonic_sort::bitonic_sort_config<1, WorkgroupSizeLog2, KeyType, ValueType, Comparator>;
+    using key_t = KeyType;
+    using value_t = ValueType;
+    using comparator_t = Comparator;
+
+    using SortConfig = subgroup::bitonic_sort_config<key_t, value_t, comparator_t>;
+
+    template<typename Key, typename Value, typename key_adaptor_t, typename value_adaptor_t>
+    static void shuffleXor(NBL_REF_ARG(pair<Key, Value>) p, uint32_t ownedIdx, uint32_t mask, NBL_REF_ARG(key_adaptor_t) keyAdaptor, NBL_REF_ARG(value_adaptor_t) valueAdaptor)
+    {
+        keyAdaptor.template set<Key>(ownedIdx, p.first);
+        valueAdaptor.template set<Value>(ownedIdx, p.second);
+
+        // Wait until all writes are done before reading - only barrier on one adaptor here
+        keyAdaptor.workgroupExecutionAndMemoryBarrier();
+
+        keyAdaptor.template get<Key>(ownedIdx ^ mask, p.first);
+        valueAdaptor.template get<Value>(ownedIdx ^ mask, p.second);
+    }
```

Comment:
@Fletterio an XOR shuffle of a struct (single accessor) within a workgroup should be commonalized somewhere cause its used in FFT and here